### PR TITLE
Multi-cursor queue components: task key and range

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Add task key & range definition and related methods

**Why?**
- Building blocks for multi-cursor queue

**How did you test it?**
- Unit test

**Potential risks**
- N/A, not used

**Is hotfix candidate?**
- No

### CODE REVIEW

File: service/history/tasks/key.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+func NewImmediateKey(taskID int64) Key {
+	return Key{
+		FireTime: DefaultFireTime,
+		TaskID:   taskID,
+	}
+}
+
+func NewKey(fireTime time.Time, taskID int64) Key {
+	return Key{
+		FireTime: fireTime,
```
Comment (wxing1292) [inline service/history/tasks/key.go:57]: add a sanity check that fireTime.UnixNano() >= 0?

File: service/history/tasks/range.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+func (r *Range) IsEmpty() bool {
+	return r.InclusiveMin.CompareTo(r.ExclusiveMax) == 0
+}
+
+func (r *Range) ContainsKey(
+	key Key,
+) bool {
+	return key.CompareTo(r.InclusiveMin) >= 0 &&
+		key.CompareTo(r.ExclusiveMax) < 0
+}
```
Comment (wxing1292) [inline service/history/tasks/range.go:61]: if InclusiveMin == ExclusiveMax == key, should this function return true or false? <- interesting Q

Comment (wxing1292) [reply service/history/tasks/range.go:61]: i believe this should return false, which is the current impl

Comment (yycptt) [reply service/history/tasks/range.go:61]: Yeah I agree. An empty range should contain no key.

File: service/history/tasks/range.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+func (r *Range) ContainsRange(
+	input Range,
+) bool {
+	containsMin := r.ContainsKey(input.InclusiveMin)
+	containsMax := r.ContainsKey(input.ExclusiveMax) || r.ExclusiveMax.CompareTo(input.ExclusiveMax) == 0
+	// Note we consider empty range = to max as contains
+	return (containsMin && containsMax) ||
+		(input.IsEmpty() && input.InclusiveMin.CompareTo(r.ExclusiveMax) == 0)
```
Comment (wxing1292) [inline service/history/tasks/range.go:70]: if r == input -> true

containsMin := r.ContainsKey(input.InclusiveMin)
containsMax := r.ContainsKey(input.ExclusiveMax) || r.ExclusiveMax.CompareTo(input.ExclusiveMax) == 0
return containsMin && containsMax

File: service/history/tasks/range.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	input Range,
+) Range {
+	if !r.CanMerge(input) {
+		panic(fmt.Sprintf("Unable to merge range %v with incoming range %v", r, input))
+	}
+
+	return NewRange(
+		MinKey(r.InclusiveMin, input.InclusiveMin),
+		MaxKey(r.ExclusiveMax, input.ExclusiveMax),
+	)
```
Comment (wxing1292) [inline service/history/tasks/range.go:106]: input.InclusiveMin can be < r.InclusiveMin
so you probably need to sort the 2 ranges first

File: service/history/tasks/range_test.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+		time.Unix(0, minKey.FireTime.UnixNano()+1),
+		minKey.TaskID,
+	)
+	r = NewRange(minKey, maxKey)
+	s.False(r.IsEmpty())
+
+	maxKey = NewKey(minKey.FireTime, minKey.TaskID+1)
+	r = NewRange(minKey, maxKey)
+	s.False(r.IsEmpty())
+}
```
Comment (wxing1292) [inline service/history/tasks/range_test.go:100]: also add a test
```
maxKey = NewKey(
	time.Unix(0, minKey.FireTime.UnixNano()+1),
	minKey.TaskID-1,
)
r = NewRange(minKey, maxKey)
s.False(r.IsEmpty())

File: service/history/tasks/range.go (hunk 4 above)
Comment (wxing1292) [reply service/history/tasks/range.go:106]: actually ignore above comment

**Review by wxing1292 — APPROVED** [review]

File: service/history/tasks/range.go (hunk 3 above)
Comment (yycptt) [reply service/history/tasks/range.go]: ^ doesn't account for the case [1,2).ContainsRange([2, 2)).
I changed to implementation to
```
r.InclusiveMin.CompareTo(input.InclusiveMin) <= 0 && r.ExclusiveMax.CompareTo(input.ExclusiveMax) >= 0
```
which should cover all cases and easier to reason about.

File: service/history/tasks/key.go (hunk 1 above)
Comment (yycptt) [reply service/history/tasks/key.go:57]: yeah I think we discussed that before, but think it's a bit risky and would first validate it on a test cluster.
The change can be found here: https://github.com/temporalio/temporal/tree/panic-invalid-task-key, I should probably rebase it on master.

**Pull request merged**